KeyDistribution_E91/ReferenceImplementation.qs

```diff
@@ -0,0 +1,199 @@
+	
+    // Task 1.1. Entangled Pairs
+    operation EntangledPairs_Reference (qsAlice : Qubit[], qsBob : Qubit[]) : Unit {
+        Fact(Length(qsAlice) == Length(qsBob), "Alice and Bob should have the same number of qubits");
+
+        for i in IndexRange(qsAlice) {
+            H(qsAlice[i]);
+            CNOT(qsAlice[i], qsBob[i]);
+        }
+    }
+
+
+    //////////////////////////////////////////////////////////////////
+    // Part II. E91 Protocol
+    //////////////////////////////////////////////////////////////////
+
+    // Task 2.1 Rotate and Measure
+    operation RotateAndMeasure_Reference (q : Qubit, rotationMultiplier: Int) : Result {
+        mutable result = Zero;
+        let rotationAngle = PI() * IntAsDouble(rotationMultiplier) / 4.0;
+
+        within {
+            Ry(rotationAngle, q);
+        } apply {
+            set result = M(q);
+        }
+
+        return result;
+    }
+
+    // Task 2.2 Measure Qubit Arrays
+    operation MeasureQubitArray_Reference (qs: Qubit[], basisList: Int[]) : (Int[], Result[]) {
+        mutable results = new Result[0];
+        mutable bases = new Int[0];
+
+        for q in qs {
+            let basis = basisList[DrawRandomInt(0,2)];
+            let res = RotateAndMeasure(q, basis);
+
+            set bases += [basis];
+            set results += [res];
+        }
+
+        return (bases, results);
+
+    }
+
+    // Task 2.3 Generate the shared key
+    function GenerateSharedKey_Reference (basesAlice: Int[], basesBob: Int[], results: Result[]) : Bool[] {
+        mutable key = new Bool[0];
+
+        for (a, b, r) in Zipped3(basesAlice, basesBob, results) {
+            if a == b {
+                set key += [ResultAsBool(r)];
+            }
+        }
+        
+        return key;
+    }
+
+    // Task 2.4 CHSH Correlation Check
+    function CorrelationCheck_Reference (basesAlice: Int[], basesBob: Int[], resAlice: Result[], resBob: Result[]) : Double {
+        mutable expectationValues = new Double[0];
+        
+        for (i, j) in [(0,1), (0, 3), (2, 1), (2, 3)] {
+            mutable sum = 0;
+            mutable counter = 0;
+
+            for idx in IndexRange(basesAlice) {
+                if basesAlice[idx] == i and basesBob[idx] == j {
+                    if resAlice[idx] == resBob[idx] {
+                        set sum += 1;
+                    } else {
+                        set sum -= 1;
+                    }
+                    set counter += 1;
+                }
+            }
+            if counter == 0 {
+                set expectationValues += [0.0];
+            } else {
+                set expectationValues += [IntAsDouble(sum)/IntAsDouble(counter)];
+            }
+        }
+
+        let s = expectationValues[0] 
+            -  expectationValues[1]
+            + expectationValues[2] 
+            + expectationValues[3]; 
+
+        return s;
+    }
+
+
+    // Task 2.5 Putting it all together 
+    @Test("QuantumSimulator")
```

Comment:
The reference shouldn't be marked as a test - if it's marked it shows up with the rest of the tests, and it can be confusing for the learner (same for T32_E91ProtocolWithEavesdropper_Reference)